### PR TITLE
load/save models, dominance metric

### PR DESCRIPTION
**Pull request by mudit1990**
@doodhwala @sarankrishna920 @rohinikapoor 
## Load/Save models
The models can now save their weights and can reuse them at a later point in time
- no of epochs has been **removed** from model init and has been put as a method arg to train method
- weights can be saved by calling model.save(). The file is saved in the models folder under the name of <run_id>_model.pth. One can choose not to save the model by setting save_model (default: True) boolean flag passed to train method
- Models init method has 2 new params, a boolean pre_trained (default: False) and load_from. If weights from file need to be used, pre_trained needs to be set to true and it reads weights from filename specified by load_from argument. The filename is searched in the models directory

## Dominance Metric

### CODE REVIEW

File: code/main.py [hunk 1]
```diff
@@ -34,7 +33,7 @@
     utils.populate_userid_mapping()
     NUM_EMAILS = 10000
 
-    model = locals()[model_name](epochs=num_epochs)
+    model = locals()[model_name](pre_trained=False, load_from='mmscdjc_model.pth')
```
Comment (mohit-surana) [inline code/main.py:36]: This is a good change.
We could also try read load_from via an optional parameter provided to the shell script
If the value is set, then we can infer that pre_trained should be set to True

File: code/metrics.py [hunk 2]
```diff
@@ -98,29 +104,70 @@ def mean_average_precision_at_k(y_true, y_pred, k=None, is_l2=False):
...
-    # (1/1 + 2/3 + 3/4 + 4/5) / 4
-    print average_precision_at_k(y2_true, y2_pred_l2, is_l2=True)
-    print average_precision_at_k(y2_true, y2_pred_prob, is_l2=False)
-
-    # (0.866 + 0.804) / 2
-    print mean_average_precision_at_k([y_true, y2_true], [y_pred_l2, y2_pred_l2], is_l2=True)
-    print mean_average_precision_at_k([y_true, y2_true], [y_pred_prob, y2_pred_prob], is_l2=False)
\ No newline at end of file
+    email_ids, embs = utils.load_user_embeddings(
+        '../important_embeddings/100usr_300em_20ep_m3/embeddings_100usr_300em_20ep_m3.pkl')
```
Comment (mohit-surana) [inline code/metrics.py]: For now, this is good.
In future, can you think of ways we can introduce this filename based on the embeddings that we just saved during the current run?

File: code/model.py [hunk 3]
```diff
@@ -20,26 +21,24 @@ def train(self, emails, w2v):
...
     def get_repr(self, identifier):
         """Given an identifier(?), get the sender or recipient's vector representation
         TODO: Consider refactoring based on requirements"""
         pass
 
+    def save(self, filename):
+        """
+        Saves the model weights to file so that they can be later loaded for inference purposes
+        """
+        torch.save(self.state_dict(), constants.MODEL_DIR + filename)
```
Comment (mohit-surana) [inline code/model.py:34]: How about a different convention to save all embeddings, models etc?
1) Create a directory in the models folder based on run id and perhaps run time params
2) All filenames won't change much and only the directory is changed
3) Code references won't have to be changed at multiple places because constants.MODEL_DIR will point to the specific model for the run.

File: code/model2.py [hunk 4]
```diff
@@ -1,110 +1,110 @@
-import numpy as np
```
Comment (mohit-surana) [inline code/model2.py:1]: You can undo this change before merging.
Why was this done though?

File: code/utils.py [hunk 5]
```diff
@@ -280,32 +283,55 @@ def extract_emb_desgn(emailids, embs):
...
+    user_desgn = load_user_designations(cat='None')
+    dom_embs = []
+    dom_y = []
+    for i in range(len(email_ids)):
+        for j in range(i+1, len(email_ids)):
+            if email_ids[i] not in user_desgn or email_ids[j] not in user_desgn or i == j:
+                continue
+            user_dg_i = user_desgn[email_ids[i]]
+            user_dg_j = user_desgn[email_ids[j]]
+            p = np.random.rand()
```
Comment (mohit-surana) [inline code/utils.py:313]: We should be able to vectorize this and `get_dominance_relation`
But since it is just around 100*100, it might not be worth prioritizing it.

File: code/model.py (hunk 3 above)
Comment (mohit-surana) [reply code/model.py:34]: In fact, you're already doing this for important embeddings.
I think doing something similar here will be better especially incase we by mistake reuse a run id
We are less likely to reuse the same hyperparameters

Comment (mohit-surana) [reply code/model.py:34]: And even if we don't give specific run ids, can we have our internal run id to be a concatenation of the user provided run id and the hyper parameters?
